cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -542,6 +542,22 @@ hash_value_type CUDA_DEVICE_CALLABLE MurmurHash3_32<double>::operator()(double c
   return this->compute_floating_point(key);
 }
 
+template <>
+hash_value_type CUDA_DEVICE_CALLABLE
+MurmurHash3_32<cudf::list_view>::operator()(cudf::list_view const& key) const
+{
+  cudf_assert(false && "List column hashing is not supported");
```

Comment:
Can this just be a static_assert? `cudf_assert` will only fail in a debug build, and only at runtime. This path should result in a compile-time error, shouldn't it?

Reply:
Yes, static_assert would be ideal.  I originally tried a static assert, but I believe the type dispatcher causes it to be referenced in the switch statement and fails the build.  I think that's why `cudf_assert` is used in [some other hash functions in this header file](https://github.com/rapidsai/cudf/blob/d4715249886522904b38236e996f369844a99e9f/cpp/include/cudf/detail/utilities/hash_functions.cuh#L158-L167) for unsupported types.

> `cudf_assert` will only fail in a debug build

That's unfortunate.  Is there a separate assertion utility that should be used here that will fire in a release build as well?

Reply:
CUDF_FAIL?

Reply:
What about a macro that simply asserts on the device in a release build without any lengthy string stuff such as function, file and line number?  Not super useful for debugging the problem, but it would prevent the code from just running and returning some bogus result without an error in case the host-side checks aren't there or working properly.  Does that still bloat the binary or harm performance in some way?

Reply:
> Does that still bloat the binary or harm performance in some way?

Yeah, the mere presence of an `assert`, even unexecuted, impacts performance. 

```
__global__ void benchmark_kernel(value_type* a, value_type* b, int n, bool do_asserts)
{
    int i = blockIdx.x * blockDim.x + threadIdx.x;
    if(i >= n)
        return;
    if(do_asserts)
    {
        assert(b[i] <= static_cast<value_type>(n));
    }
    value_type x = tanh(a[i]*sinf(b[i]));
    x = cosf(x);
    if(do_asserts)
    {
        assert(x <= 1.f);
    }
    a[i] = x;
}
```

With `do_asserts = false`:

without -DNDEBUG:
repetitions:100        median: 3.02846ms       min: 2.95219ms   p20: 2.99827ms p80: 3.05152ms  max: 5.3207ms

with -DNDEBUG:
repetitions:100        median: 2.816ms min: 2.7904ms    p20: 2.80781ms p80: 2.82726ms  max: 2.84467ms

Reply:
We should have debug builds getting tested under CI at least on a nightly/weekly basis in 0.20 which will help flag any code going down a path it shouldn't.

Reply:
As I understand it, there's nothing more to do.  The host code is already checking the types before attempting to dispatch, and these specializations using `cudf_assert` are only in place just in case somehow they get invoked errantly in practice.  The performance hit from using an actual device assert in release code is too great, so `cudf_assert` is as close as we can get, IIUC.

Reply:
Oh I thought Jake was saying just having a `cudf_assert` was expensive. Misunderstood.